File: auth_comparison_demo.rb
# frozen_string_literal: true

require "digest"
require "json"
require_relative "min0_core_forth_auth"
require_relative "image_envelope_demo"

# Public, deterministic test fixtures only. Never use these keys in deployment.
HMAC_TEST_KEY = (0...32).to_a.pack("C*").freeze
ED25519_TEST_SEED = (32...64).to_a.pack("C*").freeze
WRONG_HMAC_TEST_KEY = (0...32).to_a.reverse.pack("C*").freeze
WRONG_ED25519_TEST_SEED = ([0xA5] * 32).pack("C*").freeze

def tamper_identity(identity)
  replacement = identity[0] == "0" ? "1" : "0"
  replacement + identity[1..]
end

def benchmark_auth(identity, hmac_tag, public_key, signature)
  hmac_iterations = 2000
  ed_iterations = 300
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  hmac_iterations.times do
    Min0CoreForth::Authentication.hmac_verify(identity, HMAC_TEST_KEY, hmac_tag)
  end
  hmac_us = (Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start).fdiv(hmac_iterations * 1000)

  private_key = Min0CoreForth::Authentication.ed25519_private_from_seed(ED25519_TEST_SEED)
  start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  ed_iterations.times do
    Min0CoreForth::Authentication.ed25519_sign(identity, private_key)
  end
  ed_sign_us = (Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start).fdiv(ed_iterations * 1000)

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  ed_iterations.times do
    Min0CoreForth::Authentication.ed25519_verify(identity, public_key, signature)
  end
  ed_verify_us = (Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start).fdiv(ed_iterations * 1000)
  {
    hmac_verify_us: hmac_us.round(3),
    ed25519_sign_us: ed_sign_us.round(3),
    ed25519_verify_us: ed_verify_us.round(3),
    note: "host measurement only; not a target estimate"
  }
end

def run_auth_comparison_demo(implementation = "ruby")
  _components, envelope = build_source_image
  identity = envelope[:identity_sha256]
  tampered = tamper_identity(identity)

  hmac_tag = Min0CoreForth::Authentication.hmac_sign(identity, HMAC_TEST_KEY)
  forged_tag = Min0CoreForth::Authentication.hmac_sign(tampered, HMAC_TEST_KEY)

  private_key = Min0CoreForth::Authentication.ed25519_private_from_seed(ED25519_TEST_SEED)
  public_key = Min0CoreForth::Authentication.ed25519_public_bytes(private_key)
  wrong_public_key = Min0CoreForth::Authentication.ed25519_public_bytes(
    Min0CoreForth::Authentication.ed25519_private_from_seed(WRONG_ED25519_TEST_SEED)
  )
  signature = Min0CoreForth::Authentication.ed25519_sign(identity, private_key)
  public_object = Min0CoreForth::Authentication.ed25519_public_from_bytes(public_key)
  public_can_sign = begin
    public_object.sign(nil, Min0CoreForth::Authentication.message(identity))
    true
  rescue OpenSSL::PKey::PKeyError
    false
  end

  {
    implementation: implementation,
    identity: identity,
    message_bytes: Min0CoreForth::Authentication.message(identity).bytesize,
    hmac: {
      scheme: "hmac-sha256",
      key_id: "fixture-hmac-01",
      tag_hex: hmac_tag.unpack1("H*")
    },
    ed25519: {
      scheme: "ed25519",
      key_id: Digest::SHA256.hexdigest(public_key)[0, 16],
      public_key_hex: public_key.unpack1("H*"),
      signature_hex: signature.unpack1("H*")
    },
    sizes: {
      hmac_device_secret: HMAC_TEST_KEY.bytesize,
      hmac_tag: hmac_tag.bytesize,
      ed25519_signer_seed: ED25519_TEST_SEED.bytesize,
      ed25519_device_public: public_key.bytesize,
      ed25519_signature: signature.bytesize
    },
    verification: {
      hmac_valid: Min0CoreForth::Authentication.hmac_verify(identity, HMAC_TEST_KEY, hmac_tag),
      hmac_tampered: Min0CoreForth::Authentication.hmac_verify(tampered, HMAC_TEST_KEY, hmac_tag),
    hmac_wrong_key: Min0CoreForth::Authentication.hmac_verify(identity, WRONG_HMAC_TEST_KEY, hmac_tag),
      ed25519_valid: Min0CoreForth::Authentication.ed25519_verify(identity, public_key, signature),
      ed25519_tampered: Min0CoreForth::Authentication.ed25519_verify(tampered, public_key, signature),
      ed25519_wrong_key: Min0CoreForth::Authentication.ed25519_verify(identity, wrong_public_key, signature)
    },
    device_compromise: {
      hmac_verifier_can_forge: Min0CoreForth::Authentication.hmac_verify(tampered, HMAC_TEST_KEY, forged_tag),
      ed25519_verifier_can_forge: public_can_sign
    },
    timing: benchmark_auth(identity, hmac_tag, public_key, signature)
  }
end

puts JSON.generate(run_auth_comparison_demo) if $PROGRAM_NAME == __FILE__

File: workbench/auth_comparison_demo.py
"""Compare HMAC-SHA256 and Ed25519 for one real MIN0 CORE FORTH image identity."""

from __future__ import annotations

import hashlib
import json
import time

from min0_core_forth_auth import (
    authentication_message,
    ed25519_private_from_seed,
    ed25519_public_bytes,
    ed25519_sign,
    ed25519_verify,
    hmac_sign,
    hmac_verify,
)
from image_envelope_demo import build_source_image


# Public, deterministic test fixtures only. Never use these keys in deployment.
HMAC_TEST_KEY = bytes(range(32))
ED25519_TEST_SEED = bytes(range(32, 64))
WRONG_HMAC_TEST_KEY = bytes(reversed(range(32)))
WRONG_ED25519_TEST_SEED = bytes([0xA5] * 32)


def _tamper_identity(identity: str) -> str:
    return ("0" if identity[0] != "0" else "1") + identity[1:]


def _benchmark(identity: str, hmac_tag: bytes, public_key: bytes, signature: bytes) -> dict:
    hmac_iterations = 2000
    ed_iterations = 300
    start = time.perf_counter_ns()
    for _ in range(hmac_iterations):
        hmac_verify(identity, HMAC_TEST_KEY, hmac_tag)
    hmac_us = (time.perf_counter_ns() - start) / hmac_iterations / 1000

    private_key = ed25519_private_from_seed(ED25519_TEST_SEED)
    start = time.perf_counter_ns()
    for _ in range(ed_iterations):
        ed25519_sign(identity, private_key)
    ed_sign_us = (time.perf_counter_ns() - start) / ed_iterations / 1000

    start = time.perf_counter_ns()
    for _ in range(ed_iterations):
        ed25519_verify(identity, public_key, signature)
    ed_verify_us = (time.perf_counter_ns() - start) / ed_iterations / 1000
    return {
        "hmac_verify_us": round(hmac_us, 3),
        "ed25519_sign_us": round(ed_sign_us, 3),
        "ed25519_verify_us": round(ed_verify_us, 3),
        "note": "host measurement only; not a target estimate",
    }


def run_demo(implementation: str = "python") -> dict:
    _components, envelope = build_source_image()
    identity = envelope["identity_sha256"]
    tampered_identity = _tamper_identity(identity)

    hmac_tag = hmac_sign(identity, HMAC_TEST_KEY)
    hmac_forged_tag = hmac_sign(tampered_identity, HMAC_TEST_KEY)

    private_key = ed25519_private_from_seed(ED25519_TEST_SEED)
    public_key = ed25519_public_bytes(private_key)
    wrong_public_key = ed25519_public_bytes(
        ed25519_private_from_seed(WRONG_ED25519_TEST_SEED)
    )
    signature = ed25519_sign(identity, private_key)
    public_object_can_sign = hasattr(private_key.public_key(), "sign")

    hmac_block = {
        "scheme": "hmac-sha256",
        "key_id": "fixture-hmac-01",
        "tag_hex": hmac_tag.hex(),
    }
    ed25519_block = {
        "scheme": "ed25519",
        "key_id": hashlib.sha256(public_key).hexdigest()[:16],
        "public_key_hex": public_key.hex(),
        "signature_hex": signature.hex(),
    }
    return {
        "implementation": implementation,
        "identity": identity,
        "message_bytes": len(authentication_message(identity)),
        "hmac": hmac_block,
        "ed25519": ed25519_block,
        "sizes": {
            "hmac_device_secret": len(HMAC_TEST_KEY),
            "hmac_tag": len(hmac_tag),
            "ed25519_signer_seed": len(ED25519_TEST_SEED),
            "ed25519_device_public": len(public_key),
            "ed25519_signature": len(signature),
        },
        "verification": {
            "hmac_valid": hmac_verify(identity, HMAC_TEST_KEY, hmac_tag),
            "hmac_tampered": hmac_verify(
                tampered_identity, HMAC_TEST_KEY, hmac_tag
            ),
            "hmac_wrong_key": hmac_verify(identity, WRONG_HMAC_TEST_KEY, hmac_tag),
            "ed25519_valid": ed25519_verify(identity, public_key, signature),
            "ed25519_tampered": ed25519_verify(
                tampered_identity, public_key, signature
            ),
            "ed25519_wrong_key": ed25519_verify(
                identity, wrong_public_key, signature
            ),
        },
        "device_compromise": {
            "hmac_verifier_can_forge": hmac_verify(
                tampered_identity, HMAC_TEST_KEY, hmac_forged_tag
            ),
            "ed25519_verifier_can_forge": public_object_can_sign,
        },
        "timing": _benchmark(identity, hmac_tag, public_key, signature),
    }


if __name__ == "__main__":
    print(json.dumps(run_demo(), sort_keys=True, separators=(",", ":")))
